File: data structure and algorithms/Stack/439.ternary-expression-parser.go
package leetcode

// ParseTernary439 scans from right to left because ternary expressions are
// right-associative. When the stack top is '?', the current byte is the
// condition and the already-collapsed true/false branches are on the stack.
//
// Go data structure note: []byte is used as a stack of unresolved tokens and
// resolved one-character expressions.
//
// Time: O(n). Space: O(n).
func ParseTernary439(expression string) string {
	stack := []byte{}
	for i := len(expression) - 1; i >= 0; i-- {
		ch := expression[i]
		if len(stack) > 0 && stack[len(stack)-1] == '?' {
			stack = stack[:len(stack)-1]
			trueExpr := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			stack = stack[:len(stack)-1] // ':'
			falseExpr := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if ch == 'T' {
				stack = append(stack, trueExpr)
			} else {
				stack = append(stack, falseExpr)
			}
		} else {
			stack = append(stack, ch)
		}
	}
	return string(stack[len(stack)-1])
}

File: data structure and algorithms/Stack/439.ternary-expression-parser.java
import java.util.*;

/**
 * Algorithm:
 * Ternary expressions are right-associative, so scan from right to left. When
 * a condition is followed by a pending '?', the true and false branches on the
 * stack have already been reduced to one character each.
 *
 * Java data structures:
 * ArrayDeque<Character> is used as a stack.
 *
 * Complexity:
 * Time O(n), space O(n).
 */
class Solution {
    public String parseTernary(String expression) {
        Deque<Character> stack = new ArrayDeque<>();
        for (int i = expression.length() - 1; i >= 0; i--) {
            char ch = expression.charAt(i);
            if (!stack.isEmpty() && stack.peek() == '?') {
                stack.pop();
                char trueExpr = stack.pop();
                stack.pop();
                char falseExpr = stack.pop();
                stack.push(ch == 'T' ? trueExpr : falseExpr);
            } else {
                stack.push(ch);
            }
        }
        return String.valueOf(stack.peek());
    }
}

